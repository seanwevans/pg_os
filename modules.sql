-----------
-- MODULES
-----------

CREATE TABLE IF NOT EXISTS modules (
    id SERIAL PRIMARY KEY,
    module_name TEXT UNIQUE NOT NULL,
    loaded BOOLEAN DEFAULT FALSE,
    code TEXT,
    created_at TIMESTAMP DEFAULT now()
);

CREATE OR REPLACE FUNCTION load_module(module_name TEXT) RETURNS VOID AS $$
BEGIN
    UPDATE modules
    SET loaded = TRUE
    WHERE modules.module_name = load_module.module_name;
    -- In practice, you'd dynamically execute code or extend functionality
END;
$$ LANGUAGE plpgsql SECURITY DEFINER SET search_path = pg_catalog, pg_temp;
ALTER FUNCTION load_module(TEXT) OWNER TO pg_os_admin;

CREATE OR REPLACE FUNCTION unload_module(module_name TEXT) RETURNS VOID AS $$
BEGIN
    UPDATE modules
    SET loaded = FALSE
    WHERE modules.module_name = unload_module.module_name;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER SET search_path = pg_catalog, pg_temp;
ALTER FUNCTION unload_module(TEXT) OWNER TO pg_os_admin;
