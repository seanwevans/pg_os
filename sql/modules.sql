-- tests for load_module and unload_module
\set ECHO none
SET client_min_messages TO warning;
DROP EXTENSION IF EXISTS pg_os CASCADE;
CREATE EXTENSION pg_os;

INSERT INTO modules (module_name) VALUES ('mod1'), ('mod2');

\set ECHO queries
\set VERBOSITY terse

-- initial state
SELECT module_name, loaded FROM modules ORDER BY id;

-- load mod1
SELECT load_module('mod1');
SELECT module_name, loaded FROM modules ORDER BY id;

-- unload mod1
SELECT unload_module('mod1');
SELECT module_name, loaded FROM modules ORDER BY id;

-- load non-existent module
SELECT load_module('mod3');
SELECT module_name, loaded FROM modules ORDER BY id;

-- load mod2
SELECT load_module('mod2');
SELECT module_name, loaded FROM modules ORDER BY id;
