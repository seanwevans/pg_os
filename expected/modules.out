-- tests for load_module and unload_module
\set ECHO none
SELECT module_name, loaded FROM modules ORDER BY id;
 module_name | loaded 
-------------+--------
 mod1        | f
 mod2        | f
(2 rows)

SELECT load_module('mod1');
 load_module 
-------------
 
(1 row)

SELECT module_name, loaded FROM modules ORDER BY id;
 module_name | loaded 
-------------+--------
 mod1        | t
 mod2        | f
(2 rows)

SELECT unload_module('mod1');
 unload_module 
---------------
 
(1 row)

SELECT module_name, loaded FROM modules ORDER BY id;
 module_name | loaded 
-------------+--------
 mod1        | f
 mod2        | f
(2 rows)

SELECT load_module('mod3');
 load_module 
-------------
 
(1 row)

SELECT module_name, loaded FROM modules ORDER BY id;
 module_name | loaded 
-------------+--------
 mod1        | f
 mod2        | f
(2 rows)

SELECT load_module('mod2');
 load_module 
-------------
 
(1 row)

SELECT module_name, loaded FROM modules ORDER BY id;
 module_name | loaded 
-------------+--------
 mod1        | f
 mod2        | t
(2 rows)

